### PR TITLE
feat: add form utility components for easier form handling

### PR DESCRIPTION
**Pull request by sidarth-23**
## Summary
- The react-hook-form library was added as a dependency to enable form validation and state management.
- Added `FormItem` component to provide a context for form items
- Added `FormMessage` component to display error messages for form fields
- Added `ReactForm` object to export all form-related components
- Added `FormLabel` component to handle form labels with error styling
- Added `Form` component as a form provider
- Added `FormField` component to handle form fields with context
- Added `useFormField` hook to access form field context and state

### UPDATED ANSWERS TO COMMENTS
- The TFieldName is the best use of naming as the use of the type is a generic
- Similarly to other places where T is used

### FUTURE PLAN
- To keep this component untouched for future use in other forms, and add a abstraction for the auth pages for design consistency

### Checklist
- [x] Tested (Must)
- [ ] Test Case added
- [x] Build Successful (Must)
- [x] Sufficient Code comments added (Must)
- [ ] Attached Screenshots / Videos 
- [x] All Relevant Documents added

## Depends on
#41

### CODE REVIEW

File: packages/ui/src/partials/react-forms.tsx [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+  useFormContext,
+} from "react-hook-form"
+
+import { cn } from "../lib/tw-class-merge"
+import { Label } from "../partials/label"
+
+const Form = FormProvider
+
+type FormFieldContextValue<
+  TFieldValues extends FieldValues = FieldValues,
```
Comment (yaju09) [inline packages/ui/src/partials/react-forms.tsx]:
```suggestion
  TFieldValues extends FieldValues = FieldValues,
```
can we come up with a more meaningful name for the same?

File: packages/ui/src/partials/react-forms.tsx [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
+> = {
+  name: TName
```
Comment (yaju09) [inline packages/ui/src/partials/react-forms.tsx]:
```suggestion
  name: TName
```

can we come up with a better name?

File: packages/ui/src/partials/react-forms.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+type FormItemContextValue = {
+  id: string
+}
+
+const FormItemContext = React.createContext<FormItemContextValue>(
+  {} as FormItemContextValue
+)
+
+const FormItem = React.forwardRef<
```
Comment (yaju09) [inline packages/ui/src/partials/react-forms.tsx]:
```suggestion
const FormItem = React.forwardRef<
```
can we create separate comps for different form items like formItem, formlabel, form description etc?

**Review by yaju09 — CHANGES_REQUESTED** [review]

File: packages/ui/src/partials/index.ts [hunk 4]
```diff
@@ -3,3 +3,4 @@ export * from './card'
 export * from './checkbox'
 export * from './input'
 export * from './label'
+export * from './react-form/react-form'
```
Comment (yaju09) [inline packages/ui/src/partials/index.ts]:
```suggestion
export * from './react-form/form'
```

will it be better to rename the form comp as form?

File: packages/ui/src/partials/react-form/form-field.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+} from "react-hook-form"
+import { FormItemContext } from "./form-item"
+
+export const Form = FormProvider
+
+/** 
+ * This is an extensible type that can be used to add additional properties to the form field context
+ * */ 
+type FormFieldContextValue<
+  TFieldValues extends FieldValues = FieldValues,
```
Comment (yaju09) [inline packages/ui/src/partials/react-form/form-field.tsx:21]:
```suggestion
  TFieldValues extends FieldValues = FieldValues,
```
can we have a more meaningful name for the TFieldValues

File: packages/ui/src/partials/react-form/form-field.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
```
Comment (yaju09) [inline packages/ui/src/partials/react-form/form-field.tsx:22]:
```suggestion
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
```
more meaningful name for TName

File: packages/ui/src/partials/react-form/form-item.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import React from "react"
+import { cn } from "../../lib/tw-class-merge"
+
+type FormItemContextValue = {
+  id: string
+}
+
+export const FormItemContext = React.createContext<FormItemContextValue>(
```
Comment (yaju09) [inline packages/ui/src/partials/react-form/form-item.tsx:10]: can the context be defined in one central react function and imported from there wherever required?

File: packages/ui/src/partials/react-form/form-field.tsx [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+> = {
+  name: TName
+}
+
+/**
+ * This is a context that is used to provide the form field context to the form field
+ *
+ * @default {} as FormFieldContextValue
+ */
+const FormFieldContext = React.createContext<FormFieldContextValue>(
```
Comment (yaju09) [inline packages/ui/src/partials/react-form/form-field.tsx:32]: can the context be defined in one central react function and imported from there wherever required?

File: packages/ui/src/partials/react-form/form-field.tsx [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+/**
+ * This is a hook that is used to get the form field context
+ * 
+ * This hook should be used within the form field
+ * 
+ * @returns The form field context
+ */
+export const useFormField = () => {
```
Comment (yaju09) [inline packages/ui/src/partials/react-form/form-field.tsx:62]: can we create a separate custom in a diff comp?

**Review by yaju09 — CHANGES_REQUESTED** [review]
how will  it be to define all the contexts centrally within the react form directory and import then wherever needed ?
Also, how about defining custom hooks separate comps?

File: packages/ui/src/partials/index.ts (hunk 4 above)
Comment (sidarth-23) [reply packages/ui/src/partials/index.ts]: This can be done

File: packages/ui/src/partials/react-form/form-field.tsx (hunk 5 above)
Comment (sidarth-23) [reply packages/ui/src/partials/react-form/form-field.tsx:21]: This is a generic. Hence I think it is norm that generics start with a T in the beginning

https://wanago.io/2020/02/17/typescript-generics-discussing-naming-conventions/

File: packages/ui/src/partials/react-form/form-field.tsx (hunk 6 above)
Comment (sidarth-23) [reply packages/ui/src/partials/react-form/form-field.tsx:22]: This is a generic. Hence I think it is norm that generics start with a T in the beginning

https://wanago.io/2020/02/17/typescript-generics-discussing-naming-conventions/

File: packages/ui/src/partials/react-form/form-item.tsx (hunk 7 above)
Comment (sidarth-23) [reply packages/ui/src/partials/react-form/form-item.tsx:10]: This is not going to exported directly to be consumed. So is it worth it creating a separate file for this

File: packages/ui/src/partials/react-form/form-field.tsx (hunk 8 above)
Comment (sidarth-23) [reply packages/ui/src/partials/react-form/form-field.tsx:32]: This is not going to exported directly to be consumed. So is it worth it creating a separate file for this

File: packages/ui/src/partials/react-form/form-field.tsx (hunk 9 above)
Comment (sidarth-23) [reply packages/ui/src/partials/react-form/form-field.tsx:62]: I think I din't want to to abstract out too much. So as this is related to the field, I kept it in the same file

**Comment by sidarth-23** [comment]
Leave the build fail for now. This has to be dealt after PR#41

**Comment by sidarth-23** [comment]
Closing this to prevent merge conflict with branched out versions

**Pull request closed without merging**